### ISSUE TITLE
Biomass calculations need updated

### ISSUE THREAD

**Issue by djj4tree** [forestgeo/forum#8]
The current biomass calculations use older tropical allometries.  I am sure others have updated how they are calculating biomass on our plots. I think there should be a focused discussion on how to best keep the biomass function flexible (geographic variation in allometry) and current.

**Comment by maurolepore** [comment]
Thanks @djj4tree, 

Let me involve Sean and Gabriel in this discussion (@core).

Cheers,
Mauro

**Comment by djj4tree** [comment]
Should include Helene, Ervan and Krista as well.

**Comment by maurolepore** [comment]
@forestgeo/carbon and @teixeirak, welcome are your suggestions.

**Comment by teixeirak** [comment]
Yes, our biomass calculations desperately need updating. There has been at least one publication using tropical allometries for some temperate forests (Chisolm et al. 2013), and we need to facilitate better practices.

My lab is currently compiling best available species-specific allometries for many of the temperate sites. We have been consulting site PIs, and when they don't have specific recommendations, compiling the best fits sensu [Gonzalez-Akre et al. 2016](http://onlinelibrary.wiley.com/doi/10.1002/ecs2.1595/full), see SI). Another (more generic) option for temperate sites (at least in N. America) is to use the generic allometries of Jenkins et al. 2004, Chojnacky et al. 2014. 

For the tropics, we are using Chave et al. 2014 allometries, except for a few sites with local allometries available (Hawaii, Luquillo). We've also identified palm allometries.  

I'm happy to share all of this. If you confirm that it would be useful, I can work on getting it up on Github.

**Comment by maurolepore** [comment]
I like the direction this discussion is going. If you need to add more people to the discussion (e.g. Stuart), go ahead. If during your discussion you need anything specific from me, for example authorize a member to see a private repository, let me know.

**Comment by gabrielareto** [comment]
We need to think about the final user. How is she/he going to access/run these allometric equations?

Storing coefficients in a table is not a general solution, because it would refer only to linear models (or models that can be translated into linear models).

One option is to provide many functions directly, with all the default parameters hard-wired inside the body of the function:

Chave_2014_wet <- function(DBH, height, wood.density) { xxx xxx xxx}
Chave_2014_dry <- function(DBH, height, wood.density) { xxx xxx xxx}
Chojnacky_2014 <- function(DBH, height, wood.density) { xxx xxx xxx}

We can store somewhere, maybe within forestr as a data object, what is the appropriate function for a given site. Maybe by mirroring that separated repository with the site metadata.

I think it is important to allow the user applying the functions in any way (e.g. exclude ferns, bamboo stems, etc, apply her/his own allometries to some species, etc). That flexibility, I think, should be implemented by the user directly:

palms <- data$sp %in% vector_with_palm_spp_codes
maples <- data$sp %in% vector_with_acer_spp_codes
other <- !(palms | maples)
data$AGB[palms] <- allometry_function_for_palms(dbh = data$dbh[palms], etc)
data$AGB[maples] <- allometry_function_for_maples(dbh = data$dbh[maples], etc)
data$AGB[other] <- general_allometry_function_for_that_forest_type(dbh = data$dbh[other], etc)

I'm not sure if it is worth to create a general wrapper. We would need to make many assumptions or force the user to provide arguments in a way that may or may not be her/his preferred way.

We could create a pseudo-wrapper like colorRampPalette: a function that returns one or more functions. There we can include the best recommendation available or two or more options, if it is unclear (based on user input) what is the best reccomendation. We can link that to a (potentially large) table with species, genera, families, site characteristics, and so on. It could look like:

help_with_allometries <- function(site.id = NA, site.name = NA, forest.type = c("dry", "wet", "moist", "temperate", ...), species = NA, genus = NA, family = NA)

and depending on the combinations of input, the function could return a function or a list of functions. E.g. one enters tropical wet without any other specification and the output could look like:

list(get_AGB_for_monocots = function(xxxxxxx), get_AGB_for_dicots = function(yyyyyy))

or

list(one_option_based_on_Chave = function(xxxxxxx), another_option_based_on_X = function(yyyyyy))

**Comment by djj4tree** [comment]
I think we want a general biomass function that replaces what is currently in the R package and is friendly to PIs.  
For the big tree paper that Jim Lutz just submitted, I wrote a modified biomass calculation function that determines the "E" value (Chave etal 2014) for any tropical plot and used Chojnacky et al. 2014 temperate plots. This is not ideal but is fairly generalizable to plots.  Obviously if there are local allometries then those will supersede generalizations, but those are the exception and not the rull for most plots. 
I like the wrapper idea that Gabriel provided.

**Comment by gabrielareto** [comment]
It would be useful something that could run just based on "siteID", optionally as an attribute in the census data, and return as a default **a function that would return** the output described by @djj4tree (Chave+Chojnacky). Such a function could be expanded by adding more specific arguments and returned allometries.

xx <- function(census.data = NULL, site.id = NA, other arguments that would generate more specific allometries for palms, etc)
{
  if(!is.null(census.data)) id1 <- attr(census.data, "siteID")
  if(!is.na(site.id)) id2 <- site.id
 
  (here compare id1, id2, and return error if don't match)
  (simplify into a single id object)

  if(id == "xxxxxx") {use a site-specific allometry}

  if(!id %in% list of sites with site-specific allometries)
  {
     E.value.for.this.site <- E.values[id]
    use Chave et al. 2014
     ...
  }

  (if more specific arguments, return taxa-specific allometries etc)

  return(the selected allometric equation)
}

Where E.values is a named vector with the E values (sensu Chave et al. 2014) pre-calculated for all the sites. This should be cheap to store. A tighter link with the climatic data repository could be more accurate but probably more difficult to implement.

A function that returns a function should allow to more specific calculations, if the user wants to. E.g. using her/his own allometries for specific taxa.

Depending exclusively on objects attributes would make the forestr package less general, I think it is something to avoid.

**Comment by gabrielareto** [comment]
This is a wrapper that returns a function with the diameter:height allometric equation in a tropical tree with a E = -0.07 (e.g. Amacayacu)

`get_allometry_DH <- function(E)
{
	f <- function(D) exp(0.893 - E + 0.760 * log(D) - 0.0340 * (log(D))^2)
	attr(f, "E") <- E
	f
}
`
It works like this:

f <- get_allometry_DH(E = -0.07)
f

f(D = 30) # the expected height for a tree of that DBH

**Comment by teixeirak** [comment]
A few comments:
-Erika Gonzalez-Akre (@gonzalezeb) has compiled site specific allometries for ~10 temperate sites. 
-Valentine Herrmann (@ValentineHerr) has already written R code that applies the allometries we've compiled. I'm not familiar with its exact structure; she can provide info. You should certainly look at it before writing something new.
-We should include functions for multiple different allometries (e.g., new and old Chave; Chojnacky, sitew-specific temperate allometries). 
- Regarding E, I would not recommend linking with the climate data portal to get more exact values. The Chave et al 2014 equation was parameterized based on  specific climate data sets. Even though these are not always the best available climate data, they are still the most appropriate. As Dan mentioned, they can be calculated for sites missing from Chave's map (e.g., Hawaii sites). We have done that, as has Dan.

**Comment by ervanSTRI** [comment]
@maurolepore : For whathever reason, Helene and I haven't receive notification of this discussion. Seems that @team link does not push notifications correctly.

@ all: I have developed a home-made function that compute biomass (after having allocated wood density when available) for all tropical CTFS sites. I am using a metatable where the E index (and other useful info, i.e. plot size, way trees were measured and coded, etc) is computed in advance. I believe it is more efficient to have generic functions that source an 'updatable' table that dozen of site-specific functions. 

@teixeirak: are you aware of the FAO repository on volume and allometric models ?http://www.globallometree.org/

**Comment by maurolepore** [comment]
@ervanSTRI, RE: 
> For whathever reason, Helene and I haven't receive notification of this discussion. Seems that @Team link does not push notifications correctly.

I checked GitHub help and the team was mentioned correctly (https://github.com/blog/1121-introducing-team-mentions); Can you check that you haven't muted or ignored this thread?

From https://help.github.com/articles/unwatching-repositories/

> Ignoring Repositories: If you select Ignoring from the Watch drop-down, you won't receive any notifications, even if you're mentioned. This is not recommended, but you can use this option in the extremely rare circumstances when you really don't care about a project.

**Comment by teixeirak** [comment]
Regarding the FAO repository (Globallometree)-- Yes, we've been drawing from that.

**Comment by gabrielareto** [comment]
Related to global repositories: does anyone know about this?
BAAD: a Biomass And Allometry Database for woody plants
https://github.com/dfalster/baad
